File: src/main/java/org/taktik/icure/entities/base/Code.java
/*
 * Copyright (C) 2018 Taktik SA
 *
 * This file is part of iCureBackend.
 *
 * iCureBackend is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as published by
 * the Free Software Foundation.
 *
 * iCureBackend is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with iCureBackend.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.taktik.icure.entities.base;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.google.common.collect.ImmutableMap;
import org.taktik.icure.entities.embed.Periodicity;

import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Set;

@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonIgnoreProperties(ignoreUnknown = true)
public class Code extends StoredDocument implements CodeIdentification {
	private static final long serialVersionUID = 1L;
	public static final Map<String,String> versionsMap = ImmutableMap.of(
			"INAMI-RIZIV", "1.0"
	);

	// id = type|code|version  => this must be unique

    protected String author;

	protected Set<String> regions; //ex: be,fr
    protected List<Periodicity> periodicity;
	protected String type; //ex: ICD (type + version + code combination must be unique) (or from tags -> CD-ITEM)
    protected String code; //ex: I06.2 (or from tags -> healthcareelement). Local codes are encoded as LOCAL:SLLOCALFROMMYSOFT
    protected String version; //ex: 10. Must be lexicographically searchable

    protected Integer level; //ex: 0 = System, not to be modified by user, 1 = optional, created or modified by user

    protected java.util.Map<String, String> label; //ex: {en: Rheumatic Aortic Stenosis, fr: Sténose rhumatoïde de l'Aorte}

	@Deprecated
    protected List<String> links; //Links towards related codes (corresponds to an approximate link in qualifiedLinks)

	protected Map<LinkQualification, List<String>> qualifiedLinks; //Links towards related codes
    protected Set<CodeFlag> flags; //flags (like female only) for the code
    protected java.util.Map<String, Set<String>> searchTerms; //Extra search terms/ language

	protected String data;

    protected Map<AppendixType, String> appendices;

    protected boolean disabled;

	public static Code dataCode(String typeAndCodeAndVersion, String data) {
		Code c = new Code(typeAndCodeAndVersion);
		c.data = data;

		return c;
	}

    public Code() {
    }

    public Code(String typeAndCodeAndVersion) {
        this(typeAndCodeAndVersion.split("\\|")[0],typeAndCodeAndVersion.split("\\|")[1],typeAndCodeAndVersion.split("\\|")[2]);
    }

	public Code(String type, String code, String version) {
		this(new HashSet<>(Arrays.asList("be","fr")),type, code, version);
	}

	public Code(Set<String> regions, String type, String code, String version) {
		this(regions, type, code, version, new HashMap<>());
	}

	public Code(String region, String type, String code, String version) {
		this(Collections.singleton(region), type, code, version);
	}

    public Code(Set<String> regions, String type, String code, String version, java.util.Map<String, String> label) {
        this.regions = regions;
        this.type = type;
        this.code = code;
        this.version = version;
        this.label = label;

        this.id = type+'|'+code+'|'+version;
    }

    public String toString() {
        return this.type + ":" + this.code;
    }

    @Deprecated
    @JsonIgnore
    public String getDescrFR() {
        return label != null ? label.get("fr") : null;
    }

    @Deprecated
    public void setDescrFR(String descrFR) {
        if (label == null) { label = new HashMap<>(); }
        label.put("fr", descrFR);
    }

    @Deprecated
    @JsonIgnore
    public String getDescrNL() {
        return label != null ? label.get("nl") : null;
    }

    @Deprecated
    public void setDescrNL(String descrNL) {
        if (label == null) { label = new HashMap<>(); }
        label.put("nl", descrNL);
    }

    public String getCode() {
        return code;
    }

    public void setCode(String code) {
        this.id = type+'|'+code+'|'+version;
        this.code = code;
    }

    public Integer getLevel() {
        return level;
    }

    public void setLevel(Integer level) {
        this.level = level;
    }

    public Map<String, String> getLabel() {
        return label;
    }

    public void setLabel(Map<String, String> label) {
        this.label = label;
    }

    public String getType() {
        return type;
    }

    public void setType(String type) {
        this.id = type+'|'+code+'|'+version;
        this.type = type;
    }

	public Set<String> getRegions() {
		return regions;
	}

	public void setRegions(Set<String> regions) {
		this.regions = regions;
	}

	public String getVersion() {
		return version;
	}

	public void setVersion(String version) {
        this.id = type+'|'+code+'|'+version;
		this.version = version;
	}

    public Map<String, Set<String>> getSearchTerms() {
        return searchTerms;
    }

    public void setSearchTerms(Map<String, Set<String>> searchTerms) {
        this.searchTerms = searchTerms;
    }

    public Set<CodeFlag> getFlags() {
        return flags;
    }

    public void setFlags(Set<CodeFlag> flags) {
        this.flags = flags;
    }

    public List<String> getLinks() {
        return links;
    }

    public void setLinks(List<String> links) {
        this.links = links;
    }

	public Map<LinkQualification, List<String>> getQualifiedLinks() {
		return qualifiedLinks;
	}

	public void setQualifiedLinks(Map<LinkQualification, List<String>> qualifiedLinks) {
		this.qualifiedLinks = qualifiedLinks;
	}

	public String getData() {
		return data;
	}

	public void setData(String data) {
		this.data = data;
	}

    public String getAuthor() { return author; }

    public void setAuthor(String author) { this.author = author; }

    public Map<AppendixType, String> getAppendices() { return appendices; }

    public void setAppendices(Map<AppendixType, String> appendices) { this.appendices = appendices; }

    public boolean isDisabled() { return disabled; }

    public void setDisabled(boolean disabled) { this.disabled = disabled; }

    public List<Periodicity> getPeriodicity() { return periodicity; }

    public void setPeriodicity(List<Periodicity> periodicity) { this.periodicity = periodicity; }

    @Override
	public boolean equals(Object o) {
		if (this == o) return true;
		if (!(o instanceof Code)) return false;
		if (!super.equals(o)) return false;
		Code code1 = (Code) o;
		return Objects.equals(type, code1.type) &&
				Objects.equals(code, code1.code) &&
				Objects.equals(version, code1.version);
	}

	@Override
	public int hashCode() {
		return Objects.hash(super.hashCode(), type, code, version);
	}
}

File: src/main/kotlin/org/taktik/icure/entities/base/Code.kt
/*
 *  iCure Data Stack. Copyright (c) 2020 Taktik SA
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Affero General Public License as
 *     published by the Free Software Foundation, either version 3 of the
 *     License, or (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful, but
 *     WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *     Affero General Public License for more details.
 *
 *     You should have received a copy of the GNU Affero General Public
 *     License along with this program.  If not, see
 *     <https://www.gnu.org/licenses/>.
 */
package org.taktik.icure.entities.base

import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.fasterxml.jackson.annotation.JsonInclude
import com.fasterxml.jackson.annotation.JsonProperty
import com.github.pozo.KotlinBuilder
import org.taktik.couchdb.entity.Attachment
import org.taktik.icure.entities.embed.Periodicity
import org.taktik.icure.entities.embed.RevisionInfo
import org.taktik.icure.entities.utils.MergeUtil
import org.taktik.icure.utils.DynamicInitializer
import org.taktik.icure.utils.invoke

@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonIgnoreProperties(ignoreUnknown = true)
@KotlinBuilder
data class Code(
        @JsonProperty("_id") override val id: String,         // id = type|code|version  => this must be unique
        @JsonProperty("_rev") override val rev: String? = null,
        @JsonProperty("deleted") override val deletionDate: Long? = null,

        override val context: String? = null, //ex: When embedded the context where this code is used
        override val type: String? = null, //ex: ICD (type + version + code combination must be unique) (or from tags -> CD-ITEM)
        override val code: String? = null, //ex: I06.2 (or from tags -> healthcareelement). Local codes are encoded as LOCAL:SLLOCALFROMMYSOFT
        override val version: String? = null, //ex: 10. Must be lexicographically searchable
        override val label: Map<String, String>? = null, //ex: {en: Rheumatic Aortic Stenosis, fr: Sténose rhumatoïde de l'Aorte}

        val author: String? = null,
        val regions: Set<String> = setOf(), //ex: be,fr
        val periodicity: Set<Periodicity> = setOf(),
        val level: Int? = null, //ex: 0 = System, not to be modified by user, 1 = optional, created or modified by user
        val links: Set<String> = setOf(), //Links towards related codes (corresponds to an approximate link in qualifiedLinks)
        val qualifiedLinks: Map<LinkQualification, List<String>> = mapOf(), //Links towards related codes
        val flags: Set<CodeFlag> = setOf(), //flags (like female only) for the code
        val searchTerms: Map<String, Set<String>> = mapOf(), //Extra search terms/ language
        val data: String? = null,
        val appendices: Map<AppendixType, String> = mapOf(),
        val disabled: Boolean = false,

        @JsonProperty("_attachments") override val attachments: Map<String, Attachment>? = null,
        @JsonProperty("_revs_info") override val revisionsInfo: List<RevisionInfo>? = null,
        @JsonProperty("_conflicts") override val conflicts: List<String>? = null,
        @JsonProperty("rev_history") override val revHistory: Map<String, String>? = null

) : StoredDocument, CodeIdentification {
    companion object : DynamicInitializer<Code> {
        fun from(type: String, code: String, version: String) = Code(id = "$type|$code|$version", type = type, code = code, version = version)
    }

    fun merge(other: Code) = Code(args = this.solveConflictsWith(other))
    fun solveConflictsWith(other: Code) = super<StoredDocument>.solveConflictsWith(other) + super<CodeIdentification>.solveConflictsWith(other) + mapOf(
            "author" to (this.author ?: other.author),
            "regions" to (other.regions + this.regions),
            "periodicity" to (other.periodicity + this.periodicity),
            "level" to (this.level ?: other.level),
            "links" to (other.links + this.links),
            "qualifiedLinks" to (other.qualifiedLinks + this.qualifiedLinks),
            "flags" to (other.flags + this.flags),
            "searchTerms" to MergeUtil.mergeMapsOfSets(this.searchTerms, other.searchTerms),
            "data" to (this.data ?: other.data),
            "appendices" to (other.appendices + this.appendices),
            "disabled" to (this.disabled)
    )

    override fun withIdRev(id: String?, rev: String) = if (id != null) this.copy(id = id, rev = rev) else this.copy(rev = rev)
    override fun withDeletionDate(deletionDate: Long?) = this.copy(deletionDate = deletionDate)

    override fun normalizeIdentification(): Code {
        val parts = this.id.split("|").toTypedArray()
        return if (this.type == null || this.code == null || this.version == null) this.copy(
                type = this.type ?: parts[0],
                code = this.code ?: parts[1],
                version = this.version ?: parts[2]
        ) else this
    }
}
